pkg/proxy/metrics/metrics.go

```diff
@@ -68,9 +68,7 @@ var (
 			Subsystem: kubeProxySubsystem,
 			Name:      "network_programming_duration_seconds",
 			Help:      "In Cluster Network Programming Latency in seconds",
-			// TODO(mm4tt): Reevaluate buckets before 1.14 release.
-			// The last bucket will be [0.001s*2^20 ~= 17min, +inf)
-			Buckets: prometheus.ExponentialBuckets(0.001, 2, 20),
+			Buckets:   prometheus.ExponentialBuckets(0.015, 2, 20),
```

Comment:
Couple thoughts from me:
1. for the purpose of this SLI, I don't think anything smaller than 100ms is what we really care about it (I don't think we will see non-neglighble amount of values lower than 100ms). So I would just start with 100ms.
I'm wondering if even 100ms is something we should really care about. Maybe 250ms is also enough?

2. O(7800s) is also waaay to high that something we care about - if programming networking takes more than 2 hours it's by far non-acceptable. I don't think we care about anything bigger than small minutes - then it's obviously wrong anyway and I don't think it matters whether it was 20m or 2h.

3. Factor of 2 is quite significant - I would suggest using something smaller than 2.

So my first suggestion would be:
(0.1, 1.5, 20)  [that translates to (100ms->221s) interval]

Though ideally, I would like to have even smaller step than 1.5 (in many cases there is a difference between say 6s and 8.5 which will be treated equally in this case).

I also think that we will be targeting with SLO to a value smaller than 1m (And this will probably be somewhat round number like 30s).
So my second proposal is more custom buckets, like we did for api-call latencies:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go#L88
Here probably ranging from [0.1 or maybe even 0.25, to O(1-2m)]
So maybe sth like:
[0.25, 0.5, 1, 2, 3, 4, 5, 7.5, 10, 12.5, 15, 20, 25, 30, 35, 40, 45, 50, 60, 90]

@freehan @bowei - for thoughts

Reply:
Thanks for your insights, Wojtek, especially around numbers. 

I gave a second thought to it and I agree with you and @mm4tt, that exponential buckets with factor of 2 doesn't provide much value. For this particular case exponential buckets might not make sense at all. As you mentioned, we don't really care about value below XXXms and I agree it doesn't matter whether the value is 20m and 2h. However I think we should have some buckets for values in order of minutes. In few samples collected so far there were values above 4 minutes. From debugging perspective it'd be good to distinct between broken programming (~minutes) and broken metric/measurement (~hours). 

Going with (0.1, 1.5, 20) or getting the same buckets as api-call latency doesn't actually solve the problem, that sparked this PR. Both don't provide enough resolution for minutes latencies. What about adding few more thresholds at the end: like 120, 180, 240, 360?

Reply:
You mean after what I suggested:
[0.25, 0.5, 1, 2, 3, 4, 5, 7.5, 10, 12.5, 15, 20, 25, 30, 35, 40, 45, 50, 60, 90]
?

That sounds reasonable to me. The question is whether changing the number of buckets isn't a breaking change. Do you know that?

Reply:
Yes, I meant adding 120, 180, 240, 360 to what you propose. So in the end buckets would look like:
[0.25, 0.5, 1, 2, 3, 4, 5, 7.5, 10, 12.5, 15, 20, 25, 30, 35, 40, 45, 50, 60, 90, 120, 180, 240, 360]
Sorry for not being precise. 

This is breaking change in a similar way to changing specification of the buckets. If someone is queries specific buckets or labels or have some assumptions about their number backed in someone changing this will break their code. However for usual queries over Prometheus histograms this shouldn't have any effect (e.g. [current implementation of network programming latency SLI](https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/prometheus/manifests/prometheus-rules.yaml#L31)).

Reply:
Reading through conversation above, is the suggestion to change from exp to linear buckets? Also, as I understand it, there is really very little cost to having lots of buckets for a global metrics like this...

Reply:
What I'm suggesting isn't even linear - it's somewhat custom.

> Also, as I understand it, there is really very little cost to having lots of buckets for a global metrics like this...

So you're suggesting making even more buckets?
I remember a discussion with prometheus folks and they said that having more than 30 buckets would cause issues, but now when I recall it was for the metric that has multiple lables with many potential values.

Reply:
Yeah, for this counter, as I understand, it's would just be an array of N doubles and the update operation just (handful) of increments, so we might as well have 100 values to get the granularity we need.

Reply:
Yeah - that sounds reasonable to me (given that we don't have any metrics on that label).
When I'm thinking more about it, I actually think the values in large clusters may exceed 30s, because of long calls to iptables - we've seen those taking 20-30s in the past.

I'm still for something a bit custom though, so let's say:

Buckets: []float64{prometheus.LinearBuckets(0.25, 0.25, 20)..., prometheus.LinearBuckets(6, 1, 55)..., prometheus.LinearBuckets(65, 5, 12)...}

(roughly, requires checking what exactly it will generate).
But would give us really good granularity.

Reply:
I've slightly update what you propose, Wojtek. I removed some lower values and added more of higher ones. While I don't expect we'd ever used them in the regular work of k8s, they are useful to debug problems around the metric itself.